packages/web3-providers-ws/src/index.js

```diff
@@ -250,36 +247,42 @@ WebsocketProvider.prototype._addResponseCallback = function(payload, callback) {
  @method _timeout
  */
 WebsocketProvider.prototype._timeout = function() {
-    for(var key in this.responseCallbacks) {
-        if(this.responseCallbacks.hasOwnProperty(key)){
+    for (var key in this.responseCallbacks) {
+        if (this.responseCallbacks.hasOwnProperty(key)) {
             this.responseCallbacks[key](errors.InvalidConnection('on WS'));
             delete this.responseCallbacks[key];
         }
     }
 };
 
 
-WebsocketProvider.prototype.send = function (payload, callback) {
+WebsocketProvider.prototype.send = function(payload, callback) {
     var _this = this;
 
     if (this.connection.readyState === this.connection.CONNECTING) {
-        setTimeout(function () {
-            _this.send(payload, callback);
-        }, 10);
+        this.connection.addEventListener(
+            'open',
+            function() {
+                _this.send(payload, callback);
+            },
+            {once: true}
+        );
+
         return;
     }
 
     // try reconnect, when connection is gone
     // if(!this.connection.writable)
     //     this.connection.connect({url: this.url});
     if (this.connection.readyState !== this.connection.OPEN) {
-        console.error('connection not open on send()');
         if (typeof this.connection.onerror === 'function') {
             this.connection.onerror(new Error('connection not open'));
```

Comment:
Identical message maybe?
```suggestion
            this.connection.onerror(new Error('connection not open on send()'));
```